### ISSUE TITLE
Blue Green  ECS Deployment

### ISSUE THREAD

**Issue by arvindshekar07** [aws/aws-cdk#1559]
Hi , 
I am referencing this video  from the 2018 Re:Invent .https://www.youtube.com/watch?v=01ewawuL-IY.

1. I am struggling to get and blue green deployment setup where I have  Infrastructure stack with fargate and another stack with multiple microservice on which run on fargate.  

2. Can  some examples be provided on how to do  blue green deployment with ECR as an source? 
The confusing part here is  how to pass the docker image tag, target and task to the cdk pipeline or as a cloudformtation template to support Blue/Green Deployment.

3. It makes sense if the stack name defaults to the TS class name but then what is the use of stack name when its being provided  in the cloudformtation service.

4. How to access other stacks without having to do an TS way of logical inheritance?
5. Are there any examples where based on aws accounts  i can put a specific VPC or provision differently  ? I know there is a conditional function  for TS but I didnt find a good way to write condition based on aws account .

6. Is there any examples on how to append  the export functionality with the parameters provided to CFT at compile time? I get an TS error when  I try to an optional variable to the export name @

**Comment by rix0rrr** [comment]
Blue/Green deployment for ECS is a supported feature, but I'm not sure it's available through CloudFormation yet, and we definitely don't have support for it in the CDK as of yet. You will have to set it up by hand for now.

>  It makes sense if the stack name defaults to the TS class name but then what is the use of stack name when its being provided in the cloudformtation service.

It's not the class name, it's the second parameter that you pass to stack:

```ts
new MyStack(app, "ThisIsTheName");
```

CloudFormation orients itself around the concept of "stacks" and stacks must have a name to identify them.

> How to access other stacks without having to do an TS way of logical inheritance?

Classes (which you would combine through inheritance) give you a "kind of stack". You *instantiate* a class to get an object, which represents a *real* Stack. For example:

```ts
class MyStack extends cdk.Stack {
   // ...
}

// MyStack is not deployable yet. It might take parameters that we haven't filled in.
// To make it real, let's instantiate it:
new MyStack(app, 'Stack1', { flavor: 'banana' });
new MyStack(app, 'Stack2, { flavor: 'raspberry' });

// In fact, I can even instantiate multiple CloudFormation stacks from the same class!
// But they must have different names. 
```

This is a long way to say, once you have an object you can access attributes of it and pass those around, potentially as input arguments to another stack. It works just the same as in a regular programming language. 

Here's an example of defining an S3 bucket in one stack, and using it in another:

https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-s3/test/integ.bucket-sharing.lit.ts#L42

> Are there any examples where based on aws accounts i can put a specific VPC or provision differently ? 

I wouldn't recommend you put `if`s with literal accounts in your source code. For example, it seems better to tag your VPC and use [`VpcNetwork.importFromContext`](https://awslabs.github.io/aws-cdk/refs/_aws-cdk_aws-ec2.html#@aws-cdk/aws-ec2.VpcNetwork.importFromContext) with the [`tags`](https://awslabs.github.io/aws-cdk/refs/_aws-cdk_aws-ec2.html#@aws-cdk/aws-ec2.VpcNetworkProviderProps.tags) member. But if you must:

```ts
class MyStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    if (this.requireAccountId() !== '12345678') {
      throw new Error('You must deploy this in account 12345678!');
    }
  }
}
```

> Is there any examples on how to append the export functionality with the parameters provided to CFT at compile time? I get an TS error when I try to an optional variable to the export name @

I'm sorry, I do not understand the question. Could you explain a bit more?

**Comment by arvindshekar07** [comment]
@rix0rrr  thanks for your quick response.


> It's not the class name, it's the second parameter that you pass to stack:
> 
> ```ts
> new MyStack(app, "ThisIsTheName");
> ```
In my use case I have the same stack have same stack for UAT and PROD . In this case what would you suggest? 
- Appending the "ThisIsTheName"
- Or pass a variable?
 - Another solution 

> if (this.requireAccountId()
In the code you mentioned how do you get requireAccountId


> ```ts
> class MyStack extends cdk.Stack {
>   constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
>     super(scope, id, props);
> 
>     if (this.requireAccountId() !== '12345678') {
>       throw new Error('You must deploy this in account 12345678!');
>     }
>   }
> }
> ```
> 
> > Is there any examples on how to append the export functionality with the parameters provided to CFT at compile time? I get an TS error when I try to an optional variable to the export name @
> 
> I'm sorry, I do not understand the question. Could you explain a bit more?

I am sorry for the lack of explanation. Imagine a stack for UAT and PROD in the same account . 
I have one stack say InfastructureStack that has all the long term resources mentioned (eg ALB,fargate S3 RDS ). I have another stack called the ServiceStack that tell what ECR task and service that needs to be created in fargate and port shift them the initial service to the new ones.
Now if InfastructureStack is been referenced by the  ServiceStack how would it distinguish between UAT and PROD .
The one solution I had was to check exported resources in the parent stack and see if I can append to the end with the environment information .
![screen shot 2019-01-17 at 12 40 12 pm](https://user-images.githubusercontent.com/4720033/51340630-0cc09200-1a55-11e9-9a89-731a90d29e6a.png)

This is an attempt to do this via CDK .
I am able to do the same process via clouldformation YML definition.

**Comment by SoManyHs** [comment]
Related: https://github.com/awslabs/aws-cdk/issues/2056

**Comment by skinny85** [comment]
Since this is a duplicate of #2056 , let's consolidate by closing this issue and moving the discussion to #2056.

**Comment by konstantinj** [comment]
https://aws.amazon.com/about-aws/whats-new/2020/05/aws-cloudformation-now-supports-blue-green-deployments-for-amazon-ecs/

reopen or new issue?

**Comment by pkandasamy91** [comment]
Reopened to track support for blue-green deployments

**Comment by MrArnoldPalmer** [comment]
So looks like starting with the L2s we can add an option to TaskDefinition to enable/blue green. We can do this by outputting the hook and transform detailed in the docs here https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/blue-green.html

Some experimentation may be required to get this working properly with a nice API. It could constitute a new "pattern". I haven't seen a lot of examples of using the L1s to achieve this but it should be possible.

**Comment by MrArnoldPalmer** [comment]
The "considerations" section of the user guide has some details that look like they are going to complicate implementation.

>- You cannot include updates to resources that trigger green deployments and updates to other resources in the same stack update, as discussed in Resource Updates that Trigger ECS Blue/Green Deployments.
>- Parameters whose values are obfuscated by CloudFormation cannot be updated by the CodeDeploy service during a green deployment, and will lead to an error and stack update failure. These include:
>    - Parameters defined with the NoEcho attribute.
>    - Parameters that use dynamic references to retrieve their values from external services. For more information, see Using Dynamic References to Specify Template Values.
>- Declaring output values or importing values from other stacks is not currently supported for templates defining blue/green ECS deployments.
>- Importing existing resources is not currently supported for templates defining blue/green ECS deployments.

**Comment by MrArnoldPalmer** [comment]
As mentioned by @shamhub in #6605 there is a construct out there supporting this that may provide some guidance for implementation.

https://www.npmjs.com/package/@cloudcomponents/cdk-blue-green-container-deployment

**Comment by CONJAUMCGCG** [comment]
What is the status of Blue Green ECS Deployment for AWS CDK?

**Comment by tsykora-verimatrix** [comment]
+1 

it would be great to see support for blue/green deployments in ECS/Fargate in CDK since it's available in CloudFormation already (since 19th May)

**Comment by MrArnoldPalmer** [comment]
We are investigating how to integrate this in CDK and provide an idiomatic experience. This is not a trivial task because of the limitations of the feature in cloudformation [as mentioned above](https://github.com/aws/aws-cdk/issues/1559#issuecomment-645683141).

We are considering a number of options including custom resources and/or magically splitting out blue/green resources into their own stack and somehow hiding the limitations around export/import mechanics.

I would encourage anyone who really wants to leverage this feature now to check out third party construct libraries, e.g. https://www.npmjs.com/package/@cloudcomponents/cdk-blue-green-container-deployment

An RFC is likely required to flesh out the design of an implementation that could be included in the core library.

**Comment by jpSimkins** [comment]
I spent the last 2 weeks trying to get this to work. I did find that npm package but I really didn't want to use Lambdas for this as I don't need them when I manually set this up. So glad I found this ticket :heart: 

I will use the npm package above to get this working and will refactor once this is properly supported by the CDK. I am just glad to see I wasn't incompetent in this. I thought I was just missing something but glad to see it's not me.

Looking forward to the update and support of Fargate Blue/Green deployments in the CDK.

Is there any ETA for this? (next 6 months, over a year) Just wondering as using the npm package will suffice for now, it's not ideal.

**Comment by clareliguori** [comment]
> The "considerations" section of the user guide has some details that look like they are going to complicate implementation.
> 
> > * You cannot include updates to resources that trigger green deployments and updates to other resources in the same stack update, as discussed in Resource Updates that Trigger ECS Blue/Green Deployments.
> > * Parameters whose values are obfuscated by CloudFormation cannot be updated by the CodeDeploy service during a green deployment, and will lead to an error and stack update failure. These include:
> >   
> >   * Parameters defined with the NoEcho attribute.
> >   * Parameters that use dynamic references to retrieve their values from external services. For more information, see Using Dynamic References to Specify Template Values.
> > * Declaring output values or importing values from other stacks is not currently supported for templates defining blue/green ECS deployments.
> > * Importing existing resources is not currently supported for templates defining blue/green ECS deployments.

Hey @MrArnoldPalmer I'm not understanding why these restrictions/considerations should prevent CDK from adding support for the new Hooks section?

Ideally, I think we should unblock customers as soon as possible to be able to generate templates for blue-green deployments with the Hooks section, even at the "L1" level, and even if that template may later fail to deploy if the stack author doesn't take into account these considerations.  The considerations listed are likely to change/loosen over time if the backend implementation changes, so I'm not sure it's the right thing to build that logic into the CDK anyway and become out of date.

**Comment by skinny85** [comment]
On a related note, we recently added support for Hooks on the L1 level in #10143.

**Comment by clareliguori** [comment]
Oh awesome, thanks @skinny85!

**Comment by MrArnoldPalmer** [comment]
@clareliguori I think investigation needs to be done in order to see how these restrictions are going to impact a higher level abstraction for this in CDK. Writing a naive construct would run into these issues because users can transparently use imports/exports between stacks and dynamic references etc.

With the hooks support available we will work on getting an example of L1 usage for users to reference in order to guide users in the interim.

**Comment by srethira** [comment]
@MrArnoldPalmer Is there any timeline for example L1 usage?

**Comment by CONJAUMCGCG** [comment]
@skinny85 @MrArnoldPalmer will this change allow us to create  Amazon ECS(Blue/Green) action in Code Pipeline? 
Currently we are not able to create ecsDeploymentApplication and ecsDeploymentGroup  for blue/green deployments, which required by  Amazon ECS(Blue/Green). 
As I understand from above we will have to use Cloudformation Action Provider to deploy Blue Green template instead of having standard ECS(Blue/Green)

**Comment by martavoi** [comment]
Any updates so far?

**Comment by clareliguori** [comment]
For folks still looking for examples, I have a fully working example here of how to do the CloudFormation-based CodeDeploy blue-green deployments with CDK and CodePipeline, as well as examples of other ECS' other deployment options (non-CFN-based CodeDeploy blue-green, rolling deployments, etc).
https://github.com/aws-samples/aws-reinvent-2019-trivia-game/tree/master/trivia-backend#provision-using-infrastructure-as-code

**Comment by DmytroChymyrys** [comment]
I do understand there are some challenges and I love CDK, but why it's still not done yet. Pulumi for example already has it : https://www.pulumi.com/docs/reference/pkg/aws/codedeploy/deploymentgroup/#blue-green-deployments-with-ecs

**Comment by clareliguori** [comment]
@DmytroChymyrys The CodeDeploy deployment group setting for ECS blue-green deployments is a CloudFormation coverage gap issue.  The root issue is here: https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/483

**Comment by maheel** [comment]
Hey guys, What is the plan on releasing this?

**Comment by pgarbe** [comment]
@clareliguori @SoManyHs Since the CodeDeploy deployment group setting for ECS blue-green deployments is now supported by CloudFormation, are there any plans to continue on this?

**Comment by fasatrix** [comment]
> As mentioned by @shamhub in #6605 there is a construct out there supporting this that may provide some guidance for implementation.
> 
> https://www.npmjs.com/package/@cloudcomponents/cdk-blue-green-container-deployment

This construct is really shake and with unpredictable behavior.. The author is not even following issues and request too

**Comment by arvindshekar07** [comment]
The current implementation with aws cdk is still little hard to implement along with code deploy and another version control other than GitHub and AWS repo . It's true that I am not that active with this issue as it was raised quite a few years back and I needed a solution back then . As an alternative I had tried Terraform  which performs the deployment as I had mentioned in the description and was not locked to source code control.

 If the current issue is resolved and works better than Terraform I would switch to this as it's native to AWS and cleaner.

**Comment by dillon-odonovan** [comment]
You can't specify the launchType on the CfnService, it needs to be set on the TaskSet. If you follow the AWS sample closely you should not have issues. Currently you can't use an L2 construct for the service, you have to use the L1 CfnService. Your service will deploy on Fargate if you TaskSet and Blue Green hook are configured correctly despite not specifying Fargate as the launchType on the L1/L2 service construct. If it attempts to deploy on EC2 then you have an issue elsewhere (most likely in the Blue Green hook with the logical IDs).

Your implementation 2 looks pretty close, just remove the launchType from that construct. It also appears like you're not properly referencing the ECS Service logical ID in the Blue/Green hook - it looks like you're providing the name of the service. Try changing those two things and see if it works for you.

**Comment by NiccoloOlivieriAchille** [comment]
Hi all,

I've been working on this for weeks, but I continue to encounter problems.
Right now my code is more or less this:
```typescript
    new CfnCodeDeployBlueGreenHook(this.stack, 'CodeDeployBlueGreenHook', {
      trafficRoutingConfig: {
        type: CfnTrafficRoutingType.TIME_BASED_CANARY,
        timeBasedCanary: {
          // Shift 20% of prod traffic, then wait 15 minutes
          stepPercentage: 20,
          bakeTimeMins: 15,
        },
      },
      additionalOptions: {
        terminationWaitTimeInMinutes: 30,
      },
      // TODO: usefull to invoke a function after test traffic is routed, but before prod traffic
      lifecycleEventHooks: {},
      serviceRole: getCodeDeployRoleName(this.stage), // retrieves the correct role
      applications: [
        {
          target: {
            type: service.cfnResourceType,
            logicalId: this.stack.getLogicalId(service),
          },
          ecsAttributes: {
            taskDefinitions: [taskDefLogicalId, `${taskDefLogicalId}Green`],
            taskSets: [taskSetLogicalId, `${taskSetLogicalId}Green`],
            trafficRouting: {
              prodTrafficRoute: {
                type: CfnListener.CFN_RESOURCE_TYPE_NAME,
                logicalId: this.stack.getLogicalId(listener4prod.node.defaultChild as CfnListener),
              },
              testTrafficRoute: {
                type: CfnListener.CFN_RESOURCE_TYPE_NAME,
                logicalId: this.stack.getLogicalId(listener4test.node.defaultChild as CfnListener),
              },
              targetGroups: [
                this.stack.getLogicalId(blueTargetGroup.node.defaultChild as CfnTargetGroup),
                this.stack.getLogicalId(greenTargetGroup.node.defaultChild as CfnTargetGroup),
              ],
            },
          },
        },
      ],
    });
```

Given that it deploys perfectly the first time, as soon as I update anything inside the application code, this hook throws an error:
`Failed to create ChangeSet cdk-deploy-change-set on RedisLoggerStack: FAILED, Transform AWS::CodeDeployBlueGreen failed with: Failed to transform template. Failed to parse the "Hooks" section for the provided template. Please ensure the AWS::CodeDeploy::BlueGreen Hook matches the expected schema.`

The produced JSON is:

```json
"Hooks": {
    "CodeDeployBlueGreenHook": {
      "Type": "AWS::CodeDeploy::BlueGreen",
      "Properties": {
        "ServiceRole": "becky-code-deploy-test",
        "Applications": [
          {
            "Target": {
              "Type": "AWS::ECS::Service",
              "LogicalID": "RedisLoggerService"
            },
            "ECSAttributes": {
              "TaskDefinitions": [
                "RedisLoggerTaskDefinition73152CDA",
                "RedisLoggerTaskDefinition73152CDAGreen"
              ],
              "TaskSets": [
                "TaskSet",
                "TaskSetGreen"
              ],
              "TrafficRouting": {
                "ProdTrafficRoute": {
                  "Type": "AWS::ElasticLoadBalancingV2::Listener",
                  "LogicalID": "BeckyApplicationLoadBalancerListenerForProduction4293A970"
                },
                "TestTrafficRoute": {
                  "Type": "AWS::ElasticLoadBalancingV2::Listener",
                  "LogicalID": "BeckyApplicationLoadBalancerListenerForTest9DD6276D"
                },
                "TargetGroups": [
                  "ServiceTargetGroupBlue9182D76F",
                  "ServiceTargetGroupGreen10661C6D"
                ]
              }
            }
          }
        ],
        "TrafficRoutingConfig": {
          "Type": "TimeBasedCanary",
          "TimeBasedCanary": {
            "StepPercentage": 20,
            "BakeTimeMins": 15
          }
        },
        "AdditionalOptions": {
          "TerminationWaitTimeInMinutes": 30
        }
      }
    }
  }
```

TBH I don't know where the problem can be: every logicalId reference something real (even `TaskSet`), except from the `Green` `TaskDefinition` and `TaskSet`, right?

What else should I look for??

Thanks a lot!

**Comment by cplee** [comment]
I've create a new PR #22081 that adds support for defining CodeDeploy DeploymentGroups for ECS Services. Working on a second PR that will add support for creating new Deployments through CDK directly.

**Comment by RichardChester** [comment]
Just wanted to check I've understood right will we be able to describe BlueGreen CodeDeploy deployments using CDK, to the same extent as we can in cloud formation (I.E with target groups and services) once PR #22295 get merged in ?

**Comment by clareliguori** [comment]
@RichardChester Correct.  You can use the L1 construct today in CDK ([working example here](https://github.com/aws-samples/aws-reinvent-trivia-game/blob/master/trivia-backend/infra/codedeploy-blue-green/deployment-setup.ts#L129-L206)), but PR #22295 will make it easier through an L2 construct.

**Comment by clareliguori** [comment]
The new L2 construct is now available in version 2.50.0.  You can see an example here:
https://github.com/aws-samples/aws-reinvent-trivia-game/blob/master/trivia-backend/infra/codedeploy-blue-green/deployment-setup.ts#L123-L144

**Comment by dsdavis4** [comment]
@clareliguori  We have followed the example you posted and when we change the tag on a task definition, we see that the current revision becomes inactive. However, we do not see any revisions created in ECS console. If we create a revision manually in ECS console, we see the code deploy project begin the blue/green deployment.

Do you know what we might be missing? Is this possible without codepipeline?

**Comment by clareliguori** [comment]
> @clareliguori We have followed the example you posted and when we change the tag on a task definition, we see that the current revision becomes inactive. However, we do not see any revisions created in ECS console. If we create a revision manually in ECS console, we see the code deploy project begin the blue/green deployment.
> 
> Do you know what we might be missing? Is this possible without codepipeline?
> 
> Note: Using c#

@cplee have you seen this behavior?  I haven't had an opportunity to try to reproduce it.

**Comment by cplee** [comment]
@dsdavis4 - how are you trying to trigger new CodeDeploy deployments outside of CodePipeline? Are you using the `AWS::CodeDeploy::BlueGreen` hook to trigger deployments?

**Comment by clareliguori** [comment]
I'm able to partially reproduce this, except I do see a new revision created while the old one is marked as inactive.

Steps with my example:
1. Deploy the example, setting up the initial stacks with the initial task definition.  I see task def revision 28 as active in ECS:
```
$ aws ecs list-task-definitions --family trivia-backend
{
    "taskDefinitionArns": [
        "arn:aws:ecs:us-west-2:123456789012:task-definition/trivia-backend:28"
    ]
}
```
2. Create a new Docker image tag in the ECR repo and update the CDK code to point to that new image:
```
$ docker tag reinvent-trivia-backend 123456789012.dkr.ecr.us-west-2.amazonaws.com/reinvent-trivia-backend:v2
$ docker push 123456789012.dkr.ecr.us-west-2.amazonaws.com/reinvent-trivia-backend:v2
$ git diff
diff --git a/trivia-backend/infra/codedeploy-blue-green/deployment-setup.ts b/trivia-backend/infra/codedeploy-blue-green/deployment-setup.ts
index 645983e..f93b6ff 100644
--- a/trivia-backend/infra/codedeploy-blue-green/deployment-setup.ts
+++ b/trivia-backend/infra/codedeploy-blue-green/deployment-setup.ts
@@ -97,7 +97,7 @@ class TriviaDeploymentResourcesStack extends Stack {
     });
     taskDefinition.addContainer('Container', {
       containerName: 'web',
-      image: ecs.ContainerImage.fromEcrRepository(repo, 'latest'),
+      image: ecs.ContainerImage.fromEcrRepository(repo, 'v2'),
       portMappings: [{
         protocol: ecs.Protocol.TCP,
         containerPort: 80,
```
3. Re-deploy the CDK stack containing the task definition to create a new task definition revision:
```
$ npm run build
$ cdk deploy --app 'node deployment-setup.js' --require-approval never TriviaDeploymentResourcesTest
```
4. The old task definition revision 28 is now marked as inactive, and the new task definition revision 29 is now active.
```
$ aws ecs list-task-definitions --family trivia-backend
{
    "taskDefinitionArns": [
        "arn:aws:ecs:us-west-2:123456789012:task-definition/trivia-backend:29"
    ]
}
$ aws ecs list-task-definitions --family trivia-backend --status INACTIVE --max-items 1 --sort DESC
{
    "taskDefinitionArns": [
        "arn:aws:ecs:us-west-2:123456789012:task-definition/trivia-backend:28"
    ]
}
```

Marking the previous task definition revision as inactive is typical CloudFormation behavior, as it assumes that you will immediately update an ECS service in the same stack with the new active revision.  In this case though, because the service will be updated with the new revision outside of CloudFormation with CodeDeploy, I think we need the previous revision to stay active.  For example, if CodeDeploy needs to roll back the ECS service to the previous revision during the deployment, it will not be able to if the previous revision is inactive.

In order to keep the previous task def revision active, an UpdateReplacePolicy of "Retain" needs to be set on the task definition resource.  I have updated my example with this code:
```
    const cfnTaskDef = taskDefinition.node.defaultChild as ecs.CfnTaskDefinition;
    cfnTaskDef.applyRemovalPolicy(RemovalPolicy.RETAIN, { applyToUpdateReplacePolicy: true });
```

**Comment by dsdavis4** [comment]
@clareliguori This is the behavior we are seeing as well. It's possible we might have had a misunderstanding of the expected outcome. We expected the `cdk deploy` to set create a new task revision with the new image and deploy it, therefore automatically kicking off the code deploy build. We see now that you are manually kicking off the ecs deploy using the ecs cli, is it possible to have this all done via `cdk deploy` ?

@cplee We are not using code pipeline, we are using the L2 construct `EcsDeploymentGroup`. But it seems we might have had a misunderstanding as to what cdk deploy would do.

**Comment by clareliguori** [comment]
Correct, the construct added for this issue does not automatically start a new CodeDeploy deployment, it only sets up the CodeDeploy deployment group resource.  @cplee's PR https://github.com/aws/aws-cdk/pull/22455 covers kicking off deployments with `cdk deploy`, or you can use a [CloudFormation blue-green hook](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.CfnCodeDeployBlueGreenHook.html) instead of using CodeDeploy deployment groups.  [Example here](https://github.com/aws-samples/aws-reinvent-trivia-game/blob/master/trivia-backend/infra/cdk/ecs-service-blue-green.ts#L217).

**Comment by cplee** [comment]
For those looking for a way to start CodeDeploy deployments for ECS services within CDK, take a look at the new construct on Construct Hub to provide this functionality:

[@cdklabs/cdk-ecs-codedeploy](https://constructs.dev/packages/@cdklabs/cdk-ecs-codedeploy)

**Comment by cplee** [comment]
@dsdavis4 -  PR #22455  attempted to add `EcsDeployment` to the CDK but the guidance was to use the CDK construct hub as the mechanism to add this new capability.
